qa/rpc-tests/mvf-bu-retarget.py

```diff
@@ -11,13 +11,101 @@
 
 from test_framework.test_framework import BitcoinTestFramework
 from test_framework.util import *
+from test_framework.arith import *
 from random import randint
-import decimal
 
 # period (in blocks) from fork activation until retargeting returns to normal
 # MVF-BU TODO: Revert to 180*144
-HARDFORK_RETARGET_BLOCKS = 90*144
-FORK_BLOCK = 2017
+HARDFORK_RETARGET_BLOCKS = 90*144    # the period when retargeting returns to original
+FORK_BLOCK = 2020                    # needs to be >= 2018 to test fork difficulty reset
+POW_LIMIT = 0x207fffff
+PREFORK_BLOCKTIME = 1                # before the fork during the regtest
+ORIGINAL_DIFFADJINTERVAL = 2016      # the original difficulty adjustment interval
+STANDARD_BLOCKTIME = 600             # the standard target seconds for a block
+
+def CalculateMVFNextWorkRequired(bits, actualBlockTimeSecs, targetBlockInterval):
+    # Returns difficulty using the fork reset formula in pow.cpp:CalculateMVFNextWorkRequired()
+
+    bnPowLimit = bits2target_int(hex2bin(int2hex(POW_LIMIT))) # MVF-BU moved here
+
+    # Limit adjustment step
+    nActualTimespan = actualBlockTimeSecs
+
+    # Target by interval
+    nTargetTimespan = targetBlockInterval * STANDARD_BLOCKTIME
+
+    # permit abrupt changes for a few blocks after the fork i.e. when nTargetTimespan is < 30 minutes (MVHF-BU-DES-DIAD-5)
+    if (nTargetTimespan >= STANDARD_BLOCKTIME * 3) :
+        # prevent abrupt changes to target
+        if (nActualTimespan < nTargetTimespan/4) :
+            nActualTimespan = nTargetTimespan/4
+        if (nActualTimespan > nTargetTimespan*4) :
+            nActualTimespan = nTargetTimespan*4
+
+    # compare with debug.log
+    #print "nTargetTimespan=%d nActualTimespan=%d" % (nTargetTimespan,nActualTimespan)
+
+    # Retarget
+    bnOld = bits2target_int(hex2bin(bits))     # SetCompact
+    # MVF-BU begin: move division before multiplication
+    # at regtest difficulty, the multiplication is prone to overflowing
+    bnNew1 = bnOld / nTargetTimespan
+    bnNew2 = bnNew1 * nActualTimespan
+
+    # Test for overflow
+    if (bnNew2 / nActualTimespan != bnNew1 or bnNew2 > bnPowLimit):
+        bnNew = bnPowLimit
+    else :
+        bnNew = bnNew2
+
+    newBits = "0x%s" % bin2hex(target_int2bits(bnNew)) # GetCompact
+    nBitsReset = int(newBits,0)
+    return nBitsReset
+
+
+def CalculateMVFResetWorkRequired(bits):
+    # Returns difficulty using the fork reset formula in pow.cpp:CalculateMVFResetWorkRequired()
+
+    bnPowLimit = bits2target_int(hex2bin(int2hex(POW_LIMIT)))
+    # drop difficulty via factor
+    nDropFactor = 4
```

Comment:
Ok. The drop factor for the reset is quite simple, what's more work is documenting the change to separate the retargeting frequency and the lookback window. Our docs there are definitely outdated now and need overhaul.

I would suggest we code freeze something that we feel reasonably happy with, document it properly and ensure tests match, and then work further on developing "attack" scenarios (in the form of tests) to try to punch holes in the design. For this, I think it would help if we bring the documented design + stable code to the community and ask them to review and to formulate attacks against it.